solr/contrib/ltr/src/test/org/apache/solr/ltr/TestCacheInteractionOfPrefetchingFieldValueFeature.java

```diff
@@ -0,0 +1,530 @@
+      "storedFloatField", "storedDoubleField", "storedStrNumField", "storedStrBoolField"};
+  private static final String[] FEATURE_NAMES = new String[]{"storedIntFieldFeature", "storedLongFieldFeature",
+      "storedFloatFieldFeature", "storedDoubleFieldFeature", "storedStrNumFieldFeature", "storedStrBoolFieldFeature"};
+  private static final String MODEL_WEIGHTS = "{\"weights\":{\"storedIntFieldFeature\":0.1,\"storedLongFieldFeature\":0.1," +
+      "\"storedFloatFieldFeature\":0.1,\"storedDoubleFieldFeature\":0.1," +
+      "\"storedStrNumFieldFeature\":0.1,\"storedStrBoolFieldFeature\":0.1}}";
+
+  @Override
+  void setupSolrCluster(int numShards, int numReplicas) throws Exception {
+    // we do not want to test the scoring / ranking but the interaction with the cache
+    // because the scoring itself behaves just like the FieldValueFeature
+    // so just one shard, replica and node serve the purpose
+    setupSolrCluster(1, 1, 1);
+  }
+
+  @Test
+  public void testSimpleQuery() throws Exception {
+    ObservingPrefetchingFieldValueFeature.setBreakPrefetching(false);
+    ObservingPrefetchingFieldValueFeature.loadedFields = new HashMap<>();
+    System.setProperty(LAZY_FIELD_LOADING_CONFIG_KEY, "false");
+    // needed to clear cache because we make assertions on its content
+    reloadCollection(COLLECTION);
+
+    SolrQuery query = new SolrQuery("{!func}sub(8,field(popularity))");
+    query.setRequestHandler("/query");
+    query.setParam("rows", "8");
+    query.setFields("id,features:[fv]");
+    query.add("rq", "{!ltr model=powpularityS-model reRankDocs=8}");
+
+    QueryResponse queryResponse = solrCluster.getSolrClient().query(COLLECTION,query);
+
+    Map<String, List<List<String>>> loadedFields = ObservingPrefetchingFieldValueFeature.loadedFields;
+
+    assertEquals(loadedFields.size(), queryResponse.getResults().size());
+    for (SolrDocument doc : queryResponse.getResults()) {
+      String docId = (String) doc.getFirstValue("id");
+      if (docId.equals("1")) {
+        assertEquals(NUM_FEATURES, loadedFields.get(docId).stream()
+            // doc with id 1 has no values set for 3 of the 6 feature fields
+            .filter(fieldLoadedList -> fieldLoadedList.size() == NUM_FEATURES - 3)
+            .count());
+      } else {
+        // all the fields were loaded at once
+        assertEquals(NUM_FEATURES, loadedFields.get(docId).stream()
+            .filter(fieldLoadedList -> fieldLoadedList.size() == NUM_FEATURES)
+            .count());
+      }
+    }
+    assertTheResponse(queryResponse);
+  }
+
+  @Test
+  public void testSimpleQueryLazy() throws Exception {
+    ObservingPrefetchingFieldValueFeature.setBreakPrefetching(false);
+    ObservingPrefetchingFieldValueFeature.loadedFields = new HashMap<>();
+    System.setProperty(LAZY_FIELD_LOADING_CONFIG_KEY, "true");
+    // needed to clear cache because we make assertions on its content
+    reloadCollection(COLLECTION);
```

Comment:
I haven't yet looked in detail w.r.t. how the lazy-loading flag influences the test logic but if the lazy=false and lazy=true tests are identical (or pretty similar) then it might be worth considering to use randomisation e.g. like in https://github.com/apache/lucene-solr/blob/releases/lucene-solr/8.8.2/solr/core/src/test/org/apache/solr/search/CursorMarkTest.java#L54 i.e. before the tests `false` or `true` is chosen and that is then used for the tests. It means that a given run only covers "half" the possibilities but it's also half the code _and_ there would be no need for collection reloading (which could also slow down the test) but yes if cache clearing needs to be achieved we still would have to do that -- might adding/deleting/updating of a document have that effect indirectly or maybe there's some other way to directly clear the document cache in a test?

Reply:
Oh, the randomization is a very good idea!
With this we were able to remove half of the tests. ( 2d355c7 ) However, the reload is still needed.
I can have a look tomorrow if I find a nicer solution. At least it happens half as often now.